File: apipool_server/schemas/stats.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Stats schemas."""

from pydantic import BaseModel, Field
from datetime import datetime
from typing import Any


class StatsUsageResponse(BaseModel):
    pool_identifier: str
    period_seconds: int
    summary: dict[str, int]
    by_key: dict[str, dict[str, int]] | None = None


class StatsTimelineResponse(BaseModel):
    pool_identifier: str
    period_seconds: int
    interval: str
    data: list[dict[str, Any]]


# ── New schemas for success rate and call logs ──────────────────────────────


class KeySuccessRateItem(BaseModel):
    """Per-key success rate statistics."""
    key_identifier: str
    alias: str | None = None
    total_calls: int = 0
    success_count: int = 0
    failed_count: int = 0
    reach_limit_count: int = 0
    success_rate: float = Field(0.0, description="Success rate as a percentage (0-100)")


class SuccessRateResponse(BaseModel):
    """Pool-level success rate statistics with per-key breakdown."""
    pool_identifier: str
    period_seconds: int
    summary: KeySuccessRateItem
    by_key: list[KeySuccessRateItem] = []


class CallLogItem(BaseModel):
    """A single call log entry."""
    id: int
    key_identifier: str
    alias: str | None = None
    status: str
    finished_at: datetime


class CallLogResponse(BaseModel):
    """Paginated call log response."""
    pool_identifier: str
    period_seconds: int
    items: list[CallLogItem]
    total: int
    page: int
    page_size: int


class KeyStatsResponse(BaseModel):
    """Single key statistics across all pools it belongs to."""
    key_identifier: str
    alias: str | None = None
    period_seconds: int
    total_calls: int = 0
    success_count: int = 0
    failed_count: int = 0
    reach_limit_count: int = 0
    success_rate: float = Field(0.0, description="Success rate as a percentage (0-100)")


# ── Client stats reporting schemas ──────────────────────────────────────


class ClientCallEvent(BaseModel):
    """A single API call event reported by an SDK client."""
    key_identifier: str
    status: str
    latency: float | None = None
    method: str | None = None
    finished_at: datetime


class StatsReportRequest(BaseModel):
    """Request body for client stats reporting."""
    pool_identifier: str
    client_id: str | None = None
    events: list[ClientCallEvent]


class StatsReportResponse(BaseModel):
    """Response for client stats reporting."""
    accepted: int
    duplicates: int = 0

File: apipool_server/frontend/src/api/stats.ts
import http from './index'

// ── Types ────────────────────────────────────────────────────────────────

export interface KeySuccessRateItem {
  key_identifier: string
  alias: string | null
  total_calls: number
  success_count: number
  failed_count: number
  reach_limit_count: number
  success_rate: number
}

export interface SuccessRateResponse {
  pool_identifier: string
  period_seconds: number
  summary: KeySuccessRateItem
  by_key: KeySuccessRateItem[]
}

export interface CallLogItem {
  id: number
  key_identifier: string
  alias: string | null
  status: string
  finished_at: string
}

export interface CallLogResponse {
  pool_identifier: string
  period_seconds: number
  items: CallLogItem[]
  total: number
  page: number
  page_size: number
}

export interface KeyStatsResponse {
  key_identifier: string
  alias: string | null
  period_seconds: number
  total_calls: number
  success_count: number
  failed_count: number
  reach_limit_count: number
  success_rate: number
}

export interface SuccessRateParams {
  seconds?: number
}

export interface CallLogParams {
  seconds?: number
  key_identifier?: string
  status?: 'success' | 'failed' | 'reach_limit'
  page?: number
  page_size?: number
}

export interface KeyStatsParams {
  seconds?: number
}

// ── API Functions ────────────────────────────────────────────────────────

export function getSuccessRate(poolIdentifier: string, params?: SuccessRateParams) {
  return http.get<SuccessRateResponse>(`/stats/${poolIdentifier}/success-rate`, { params })
}

export function getCallLogs(poolIdentifier: string, params?: CallLogParams) {
  return http.get<CallLogResponse>(`/stats/${poolIdentifier}/logs`, { params })
}

export function getKeyStats(keyIdentifier: string, params?: KeyStatsParams) {
  return http.get<KeyStatsResponse>(`/keys/${keyIdentifier}/stats`, { params })
}
